### ISSUE TITLE
Working with TensorMap

### ISSUE THREAD

**Issue by Bbllaaddee** [romeric/Fastor#132]
Hey there!

I'm trying to write the following piece of code:
```c++
#include <iostream>
#include <array>
#include <Fastor/Fastor.h>

using namespace Fastor;

int main()
{
	std::array<double,3> a = {1,1,1};
	TensorMap<double,3> at(a.data());
	
	std::array<double,3> b = {2,2,2};
	TensorMap<double,3> bt(b.data());
	
	Tensor<double,3> qwe = bt + at;
	std::cout << qwe << std::endl;
}
```
Unfortunately, the compiler gives me the following error:
```
In file included from lib/Fastor/expressions/linalg_ops/binary_matmul_op.h:10,
                 from lib/Fastor/expressions/linalg_ops/linalg_ops.h:6,
                 from lib/Fastor/tensor/Tensor.h:11,
                 from lib/Fastor/Fastor.h:19,
                 from main.cpp:3:
lib/Fastor/tensor/TensorTraits.h: In instantiation of ‘struct Fastor::scalar_type_finder<Fastor::BinaryAddOp<Fastor::TensorMap<double, 3>, Fastor::TensorMap<double, 3>, 1> >’:
lib/Fastor/expressions/binary_ops/binary_arithmetic_ops.h:215:1:   required from ‘struct Fastor::BinaryAddOp<Fastor::TensorMap<double, 3>, Fastor::TensorMap<double, 3>, 1>’
main.cpp:15:30:   required from here
lib/Fastor/tensor/TensorTraits.h:45:11: error: ambiguous template instantiation for ‘struct Fastor::scalar_type_finder<Fastor::TensorMap<double, 3> >’
   45 |     using type = typename std::conditional<std::is_arithmetic<TLhs>::value,
      |           ^~~~
lib/Fastor/tensor/TensorTraits.h:40:8: note: candidates are: ‘template<template<class, long unsigned int <anonymous> > class UnaryExpr, class Expr, long unsigned int DIMS> struct Fastor::scalar_type_finder<UnaryExpr<Expr, DIMS> > [with UnaryExpr = Fastor::TensorMap; Expr = double; long unsigned int DIMS = 3]’
   40 | struct scalar_type_finder<UnaryExpr<Expr,DIMS>> {
      |        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
lib/Fastor/tensor/TensorTraits.h:63:8: note:                 ‘template<class T, long unsigned int ...Rest> struct Fastor::scalar_type_finder<Fastor::TensorMap<T, Rest ...> > [with T = double; long unsigned int ...Rest = {3}]’
   63 | struct scalar_type_finder<TensorMap<T,Rest...>> {
      |        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
lib/Fastor/tensor/TensorTraits.h:45:11: error: invalid use of incomplete type ‘struct Fastor::scalar_type_finder<Fastor::TensorMap<double, 3> >’
   45 |     using type = typename std::conditional<std::is_arithmetic<TLhs>::value,
      |           ^~~~
lib/Fastor/tensor/TensorTraits.h:26:8: note: declaration of ‘struct Fastor::scalar_type_finder<Fastor::TensorMap<double, 3> >’
   26 | struct scalar_type_finder {
      |        ^~~~~~~~~~~~~~~~~~
In file included from lib/Fastor/tensor/Tensor.h:190,
                 from lib/Fastor/Fastor.h:19,
                 from main.cpp:3:
lib/Fastor/tensor/TensorAssignment.h: In instantiation of ‘void Fastor::trivial_assign(Fastor::AbstractTensor<Derived, DIM>&, const Fastor::AbstractTensor<Derived1, DIM1>&) [with Derived = Fastor::Tensor<double, 3>; long unsigned int DIM = 1; OtherDerived = Fastor::BinaryAddOp<Fastor::TensorMap<double, 3>, Fastor::TensorMap<double, 3>, 1>; long unsigned int OtherDIM = 1]’:
lib/Fastor/expressions/binary_ops/binary_arithmetic_assignment.h:107:1:   required from ‘void Fastor::assign(Fastor::AbstractTensor<Derived, DIM>&, const Fastor::BinaryAddOp<TLhs, TRhs, OtherDIM>&) [with Derived = Fastor::Tensor<double, 3>; long unsigned int DIM = 1; TLhs = Fastor::TensorMap<double, 3>; TRhs = Fastor::TensorMap<double, 3>; long unsigned int OtherDIM = 1; typename std::enable_if<(((! is_primitive_v_<TLhs>) && (! is_primitive_v_<TRhs>)) && (!(requires_evaluation_v<TLhs> || requires_evaluation_v<TRhs>))), bool>::type <anonymous> = false]’
lib/Fastor/tensor/Tensor.h:99:15:   required from ‘Fastor::Tensor<T, Rest>::Tensor(const Fastor::AbstractTensor<Derived, DIMS>&) [with Derived = Fastor::BinaryAddOp<Fastor::TensorMap<double, 3>, Fastor::TensorMap<double, 3>, 1>; long unsigned int DIMS = 1; typename std::enable_if<((((! has_tensor_view_v<Derived>) && (! has_tensor_fixed_view_2d_v<Derived>)) && (! has_tensor_fixed_view_nd_v<Derived>)) || (DIMS != sizeof... (Rest))), bool>::type <anonymous> = false; T = double; long unsigned int ...Rest = {3}]’
main.cpp:15:30:   required from here
lib/Fastor/tensor/TensorAssignment.h:19:26: error: ‘const struct Fastor::BinaryAddOp<Fastor::TensorMap<double, 3>, Fastor::TensorMap<double, 3>, 1>’ has no member named ‘eval’
   19 |             src.template eval<T>(i).store(&_data[i], FASTOR_ALIGNED);
      |             ~~~~~~~~~~~~~^~~~~~~
lib/Fastor/tensor/TensorAssignment.h:22:37: error: ‘const struct Fastor::BinaryAddOp<Fastor::TensorMap<double, 3>, Fastor::TensorMap<double, 3>, 1>’ has no member named ‘eval_s’
   22 |             _data[i] = src.template eval_s<T>(i);
      |                        ~~~~~~~~~~~~~^~~~~~~~~
make: *** [Makefile:2: compile] Error 1
```
Is there any way of working with 'TensorMap's in this way?

Thanks!

**Comment by matthiasneuner** [comment]
Hi, your example looks good any works perfectly for me on g++ 10.2.0. Which compiler are you using?

**Comment by Bbllaaddee** [comment]
I'm using g++ 9.3.0

I tried to compile it again and the following happened: it DOES compile with just bare g++ main.cpp, however my error appears only with g++ -std=c++17 main.cpp

Any other previous std, like -std=c++14 or -std=c++11 work.

**Comment by Bbllaaddee** [comment]
So, is there any way of compiling it under c++17 standard?)
Thx

**Comment by romeric** [comment]
Fixed here 1b30505. Please download/clone the latest copy of Fastor.

Thanks for reporting.

**Comment by Bbllaaddee** [comment]
Wow, thanks!
It does compile and works.

Unfortunately, I guess I found another bug or smth regarding this code:

```c++
#include <iostream>
#include <array>
#include <Fastor/Fastor.h>

using namespace Fastor;

int main()
{
	std::array<double,3> a = {1,1,1};
	TensorMap<double,3> a_map(a.data());
			
	std::array<double,3> b = {2,2,2};
	TensorMap<double,3> b_map(b.data());
					
	Tensor<double,3> a_tensor = {1,1,1};
	Tensor<double,3> b_tensor= {2,2,2};
	
	Tensor<double,3,3> qwe1 = outer(a_tensor, b_tensor); // works
	Tensor<double,3,3> qwe2 = outer(a_map, b_map); // compilation error

	std::cout << qwe1 << std::endl;
	std::cout << qwe2 << std::endl;
}
```
with the following g++ 9.3.0 error:
```
In file included from lib/Fastor/tensor_algebra/einsum.h:13,
                 from lib/Fastor/Fastor.h:24,
                 from main.cpp:3:
lib/Fastor/tensor_algebra/outerproduct.h: In instantiation of ‘auto Fastor::outer(const AbstractTensorType0&, const AbstractTensorType1&, const AbstractTensorTypes& ...) [with AbstractTensorType0 = Fastor::TensorMap<double, 3>; AbstractTensorType1 = Fastor::TensorMap<double, 3>; AbstractTensorTypes = {}]’:
main.cpp:19:46:   required from here
lib/Fastor/tensor_algebra/outerproduct.h:105:27: error: use of ‘auto Fastor::outer(const AbstractTensorType0&, const AbstractTensorType1&, const AbstractTensorTypes& ...) [with AbstractTensorType0 = Fastor::TensorMap<double, 3>; AbstractTensorType1 = Fastor::TensorMap<double, 3>; AbstractTensorTypes = {}]’ before deduction of ‘auto’
  105 |     const auto res = outer(a,b);
      |                      ~~~~~^~~~~
main.cpp: In function ‘int main()’:
main.cpp:19:33: error: conversion from ‘void’ to non-scalar type ‘Fastor::Tensor<double, 3, 3>’ requested
   19 |  Tensor<double,3,3> qwe2 = outer(a_map, b_map); // compilation error
      |                            ~~~~~^~~~~~~~~~~~~~
make: *** [Makefile:2: compile] Error 1

```

**Comment by romeric** [comment]
Fixed 26fc041. Should be available when you get a new copy of Fastor. In the future please open a new issue, so we can track.

Thanks
